azurerm/internal/services/mssql/resource_arm_mssql_virtual_machine.go

```diff
@@ -0,0 +1,503 @@
+							Required:     true,
+							ForceNew:     true,
+							Sensitive:    true,
+							ValidateFunc: validate.URLIsHTTPS,
+						},
+
+						"service_principal_name": {
+							Type:         schema.TypeString,
+							Required:     true,
+							ForceNew:     true,
+							Sensitive:    true,
+							ValidateFunc: validate.NoEmptyStrings,
+						},
+
+						"service_principal_secret": {
+							Type:         schema.TypeString,
+							Required:     true,
+							ForceNew:     true,
+							Sensitive:    true,
+							ValidateFunc: validate.NoEmptyStrings,
+						},
+					},
+				},
+			},
+
+			"server_configuration": {
+				Type:     schema.TypeList,
+				Optional: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"is_r_services_enabled": {
+							Type:     schema.TypeBool,
+							Optional: true,
+						},
+
+						"sql_connectivity_type": {
+							Type:     schema.TypeString,
+							Optional: true,
+							Default:  string(sqlvirtualmachine.PRIVATE),
+							ValidateFunc: validation.StringInSlice([]string{
+								string(sqlvirtualmachine.LOCAL),
+								string(sqlvirtualmachine.PRIVATE),
+								string(sqlvirtualmachine.PUBLIC),
+							}, false),
+						},
+
+						"sql_connectivity_port": {
+							Type:         schema.TypeInt,
+							Optional:     true,
+							Default:      1433,
+							ValidateFunc: validation.IntBetween(1024, 65535),
+						},
+
+						"sql_connectivity_update_password": {
+							Type:         schema.TypeString,
+							Optional:     true,
+							Sensitive:    true,
+							ValidateFunc: validate.NoEmptyStrings,
+						},
+
+						"sql_connectivity_update_username": {
+							Type:         schema.TypeString,
+							Optional:     true,
+							Sensitive:    true,
+							ValidateFunc: validate.NoEmptyStrings,
+						},
+					},
+				},
+			},
+
+			"tags": tags.Schema(),
+		},
+	}
+}
+
+func resourceArmMsSqlVirtualMachineCreateUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).MSSQL.SQLVirtualMachinesClient
+	ctx, cancel := timeouts.ForCreate(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	vmId := d.Get("virtual_machine_id").(string)
+	id, err := compute.ParseVirtualMachineID(vmId)
+	if err != nil {
+		return err
+	}
+	name := id.Name
+	resourceGroupName := id.ResourceGroup
+
+	if features.ShouldResourcesBeImported() && d.IsNewResource() {
+		existing, err := client.Get(ctx, resourceGroupName, name, "*")
+		if err != nil {
+			if !utils.ResponseWasNotFound(existing.Response) {
+				return fmt.Errorf("Error checking for present of existing Sql Virtual Machine (Sql Virtual Machine Name %q / Resource Group %q): %+v", name, resourceGroupName, err)
+			}
+		}
+		if existing.ID != nil && *existing.ID != "" {
+			return tf.ImportAsExistsError("azurerm_mssql_virtual_machine", *existing.ID)
+		}
+	}
+
+	// get location from vm
+	vmclient := meta.(*clients.Client).Compute.VMClient
```

Comment:
could we move this line up so all clients are defined at the top of the funtion?